### ISSUE TITLE
When using .fit() to train a binary classification model, the metrics (i.e. loss, accuracy) shown in the history log of .fit() for training set is different from the result from .evaluate()

### ISSUE THREAD

**Issue by JoeSu666** [tensorflow/tensorflow#52207]
I'm using the tf.keras.Model.fit() to train a binary classification model for images. I found that the metrics like accuracy shown in the final epoch's log of .fit() can be 1.0, while the accuracy shown in model.evaluate(ds_train) for training set can be only 0.71. The loss and any other metrics are also different. However, the results for validation set are perfectly consistent for .fit and .evaluate. And also, for multiclass classification training, this issue never happens.

The issue has happened multiple times for different projects. Please let me know if anything wrong happened.

Here is a sample code for my binary classification training:

def load_image(img_path,size = (32,32)):
    label = tf.constant([1],tf.int8) if tf.strings.regex_full_match(img_path,".*automobile.*") \
            else tf.constant([0],tf.int8)
    img = tf.io.read_file(img_path)
    img = tf.image.decode_jpeg(img) 
    img = tf.image.resize(img,size)/255.0
    return(img,label)

ds_train = tf.data.Dataset.list_files("./data/cifar2/train/*/*.jpg") \
           .map(load_image, num_parallel_calls=tf.data.experimental.AUTOTUNE) \
           .shuffle(buffer_size = 1000).batch(BATCH_SIZE) \
           .prefetch(tf.data.experimental.AUTOTUNE)  

ds_test = tf.data.Dataset.list_files("./data/cifar2/test/*/*.jpg") \
           .map(load_image, num_parallel_calls=tf.data.experimental.AUTOTUNE) \
           .batch(BATCH_SIZE) \
           .prefetch(tf.data.experimental.AUTOTUNE)  
tf.keras.backend.clear_session() 

inputs = layers.Input(shape=(32,32,3))
x = layers.Conv2D(32,kernel_size=(3,3))(inputs)
x = layers.MaxPool2D()(x)
x = layers.Conv2D(64,kernel_size=(5,5))(x)
x = layers.MaxPool2D()(x)
x = layers.Dropout(rate=0.1)(x)
x = layers.Flatten()(x)
x = layers.Dense(32,activation='relu')(x)
outputs = layers.Dense(1,activation = 'sigmoid')(x)

model = models.Model(inputs = inputs,outputs = outputs)

model.compile(
        optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
        loss=tf.keras.losses.binary_crossentropy,
        metrics=["accuracy"]
    )

history = model.fit(ds_train,epochs= 5,validation_data=ds_test)

*******************************************************************
after running for .fit():
Epoch 5/5
100/100 [==============================] - 1s 7ms/step - loss: 0.2098 - accuracy: 0.9154 - val_loss: 0.2171 - val_accuracy: 0.9080

for .evaluate() after training:
model.evaluate(ds_train)

100/100 [==============================] - 0s 5ms/step - loss: 0.1968 - accuracy: 0.9195

The accuracy can be more different for some other projects.

**Comment by tilakrayal** [comment]
@JoeSu666 ,
In order to expedite the trouble-shooting process, could you please provide a complete code  and the TensorFlow version you are using.

**Comment by JoeSu666** [comment]
`def load_image(img_path,size = (32,32)):
  label = tf.constant([1],tf.int8) if tf.strings.regex_full_match(img_path,".automobile.")
  else tf.constant([0],tf.int8)
  img = tf.io.read_file(img_path)
  img = tf.image.decode_jpeg(img)
  img = tf.image.resize(img,size)/255.0
  return(img,label)

ds_train = tf.data.Dataset.list_files("./data/cifar2/train//.jpg")
.map(load_image, num_parallel_calls=tf.data.experimental.AUTOTUNE)
.shuffle(buffer_size = 1000).batch(BATCH_SIZE)
.prefetch(tf.data.experimental.AUTOTUNE)

ds_test = tf.data.Dataset.list_files("./data/cifar2/test//.jpg")
.map(load_image, num_parallel_calls=tf.data.experimental.AUTOTUNE)
.batch(BATCH_SIZE)
.prefetch(tf.data.experimental.AUTOTUNE)
tf.keras.backend.clear_session()

inputs = layers.Input(shape=(32,32,3))
x = layers.Conv2D(32,kernel_size=(3,3))(inputs)
x = layers.MaxPool2D()(x)
x = layers.Conv2D(64,kernel_size=(5,5))(x)
x = layers.MaxPool2D()(x)
x = layers.Dropout(rate=0.1)(x)
x = layers.Flatten()(x)
x = layers.Dense(32,activation='relu')(x)
outputs = layers.Dense(1,activation = 'sigmoid')(x)

model = models.Model(inputs = inputs,outputs = outputs)

model.compile(
optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
loss=tf.keras.losses.binary_crossentropy,
metrics=["accuracy"]
)

history = model.fit(ds_train,epochs= 5,validation_data=ds_test)
model.evaluate(ds_train)
`
This is one of the code that yield the issue. The version I'm using is tf2.2.0

**Comment by tilakrayal** [comment]
@JoeSu666 ,
Can you please try to execute your code in latest stable version 2.6.0 and let us know if the issue still persists.Thanks!
